### PR TITLE
[C API] Add more C API for infer request

### PR DESCRIPTION
**Pull request by zhaixuejun1993**
### Details:
 - Add more C APIs about infer request
 - Add unite test for those APIs

### Tickets:
 - 89486

### CODE REVIEW

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 1]
```diff
@@ -69,16 +70,116 @@ typedef struct {
 OPENVINO_C_API(ov_status_e)
 ov_infer_request_set_tensor(ov_infer_request_t* infer_request, const char* tensor_name, const ov_tensor_t* tensor);
 
+/**
+ * @brief Sets an input/output tensor to infer on.
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:74]: update brief

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 2]
```diff
@@ -69,16 +70,116 @@ typedef struct {
...
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param port  Port of the input or output tensor.
+ * @param tensor Reference to the tensor.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_tensor_by_node(ov_infer_request_t* infer_request, const ov_output_node_t* port, const ov_tensor_t* tensor);
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:82]: Which will be better? by_node or by_port or other else?

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 3]
```diff
@@ -69,16 +70,116 @@ typedef struct {
...
+
+/**
+ * @brief Sets an input/output tensor to infer on.
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:85]: update description of brief

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 4]
```diff
@@ -69,16 +70,116 @@ typedef struct {
...
+ * @brief Sets a batch of tensors for input data to infer by tensor name.
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param tensor_name  Name of the input tensor.
+ * @param tensors Input tensors for batched infer request. The type of each tensor must match the model
+ * input element type and shape (except batch dimension). Total size of tensors must match the input size.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_tensors(ov_infer_request_t* infer_request, const char* tensor_name, const ov_tensor_list_t* tensors);
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:105]: which will be better? *_set_tensors or *_set_tensor_list or *_set_batch_tensors ? To align C++ method, it should be *_set_tensors, but user is easy to make mistake for *_set_tensor() and *_set_tensors().

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 5]
```diff
@@ -69,16 +70,116 @@ typedef struct {
...
  * @param infer_request A pointer to the ov_infer_request_t.
+ * @param idx Index of the input port. If @p idx is greater than the number of model inputs, an exception is thrown.
+ * @param tensor Reference to the tensor.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_input_tensor_by_index(ov_infer_request_t* infer_request, size_t idx, const ov_tensor_t* tensor);
+
+/**
+ * @brief Sets an input tensor for model with single input to infer on.
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:131]: Solve syntax error.

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 6]
```diff
@@ -69,16 +70,116 @@ typedef struct {
...
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param tensor Reference to the tensor.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_input_tensor(ov_infer_request_t* infer_request, const ov_tensor_t* tensor);
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:138]: It should be default interface for set input tensor.

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 7]
```diff
@@ -100,7 +244,17 @@ ov_infer_request_get_tensor(const ov_infer_request_t* infer_request, const char*
...
+
+/**
+ * @brief Gets an output tensor to infer on from single input model.
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param tensor Reference to the tensor.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_get_output_tensor(const ov_infer_request_t* infer_request, ov_tensor_t** tensor);
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:257]: It is peer method of *_get_input_tensor()

File: src/bindings/c/include/openvino/c/ov_tensor.h [hunk 8]
```diff
@@ -24,6 +24,38 @@ typedef struct ov_tensor ov_tensor_t;
...
+/**
+ * @brief Add a tensor to a list
+ * @ingroup tensor
+ * @param tensor A point to the tensor will be added to tensor list
+ * @param tensors A point to tensor list
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_tensor_list_add(ov_tensor_list_t* tensors, const ov_tensor_t* tensor);
+
```
Comment (riverlijunjie) [inline src/bindings/c/include/openvino/c/ov_tensor.h:51]: Do we need a helper function to get some info of ov_tensor_list_t, such as get tensor number?

File: src/bindings/c/src/ov_infer_request.cpp [hunk 9]
```diff
@@ -128,6 +348,19 @@ ov_status_e ov_infer_request_wait(ov_infer_request_t* infer_request) {
     return ov_status_e::OK;
 }
 
+ov_status_e ov_infer_request_wait_for(ov_infer_request_t* infer_request, const int timeout) {
```
Comment (riverlijunjie) [inline src/bindings/c/src/ov_infer_request.cpp:351]: int64_t will be better for `timeout`

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 10]
```diff
@@ -69,16 +70,116 @@ typedef struct {
...
+ * @brief Sets a batch of tensors for input data to infer by input port.
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param port  Port of the input tensor.
+ * @param tensors Input tensors for batched infer request. The type of each tensor must match the model
+ * input element type and shape (except batch dimension). Total size of tensors must match the input size.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_tensors_by_const_node(ov_infer_request_t* infer_request, const ov_output_const_node_t* port, const ov_tensor_list_t* tensors);
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:117]: The same

File: src/bindings/c/include/openvino/c/ov_tensor.h [hunk 11]
```diff
@@ -24,6 +24,38 @@ typedef struct ov_tensor ov_tensor_t;
  * @{
  */
 
+/**
+ * @struct ov_tensor_list_t
+ * @brief A list of tensors
+ */
+typedef struct ov_tensor_list ov_tensor_list_t;
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_tensor.h:31]: Why do we need to have tensor list hidden structure?
Can we use C array instead?

File: src/bindings/c/include/openvino/c/ov_tensor.h [hunk 12]
```diff
@@ -24,6 +24,38 @@ typedef struct ov_tensor ov_tensor_t;
...
+
+/**
+ * @brief Add a tensor to a list
+ * @ingroup tensor
+ * @param tensor A point to the tensor will be added to tensor list
+ * @param tensors A point to tensor list
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_tensor_list_add(ov_tensor_list_t* tensors, const ov_tensor_t* tensor);
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_tensor.h:50]: I don't believe honestly that somebody will add tensors to the list. I think we can provide more C aligned API and allow user to use C arrays.

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 2 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:82]: ok, let use _by_port.

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 4 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:105]: ok, we keep it as *_set_tensors.

File: src/bindings/c/include/openvino/c/ov_tensor.h (hunk 11 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_tensor.h:31]: ov_tensor is opaque handle of ov::Tensor, so if apply ov_tensor into C array, the array must be pre-allocated and the size is fixed. Will it make sense?

File: src/bindings/c/include/openvino/c/ov_tensor.h (hunk 12 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_tensor.h:50]: If the tensor list size is a fixed value, it is fine to use such C array. But if size of tensor list is dynamic, C array cannot support.
@ilyachur How do you think?

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 2 above)
Comment (zhaixuejun1993) [reply src/bindings/c/include/openvino/c/ov_infer_request.h]: done

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 3 above)
Comment (zhaixuejun1993) [reply src/bindings/c/include/openvino/c/ov_infer_request.h]: done

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 4 above)
Comment (zhaixuejun1993) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:105]: done

File: src/bindings/c/include/openvino/c/ov_tensor.h (hunk 12 above)
Comment (ilyachur) [reply src/bindings/c/include/openvino/c/ov_tensor.h:50]: @riverlijunjie  do you have an example where model have a dynamic set of inputs or outputs?
Different models can have different numbers of inputs, but the size of inputs and outputs at the current moment is always constant and cannot be changed in the runtime after the compilation of the model.
I think it is not a big deal to allocate the memory for all input/output tensors from the model before the filling tensors.

Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_tensor.h:50]: @ilyachur I got your point and agree with you. Considered that the tensor list size always is fixed, we don't need use a dynamic size tensor list definition.
@zhaixuejun1993  let's follow Ilya's idea :)

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 13]
```diff
@@ -70,37 +71,204 @@ OPENVINO_C_API(ov_status_e)
...
+ * with output (or any other non-input node), an exception is thrown.
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param tensor_name Name of the input tensor.
+ * @param tensors Input tensors for batched infer request. The type of each tensor must match the model
+ * input element type and shape (except batch dimension). Total size of tensors must match the input size.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_tensors(ov_infer_request_t* infer_request, const char* tensor_name, const ov_tensor_list_t* tensors);
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:108]: Looks like we can pass tensor_list by value

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 14]
```diff
@@ -70,37 +71,204 @@ OPENVINO_C_API(ov_status_e)
...
+ * with output (or any other non-input node), an exception is thrown.
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param port Const port of the input tensor.
+ * @param tensors Input tensors for batched infer request. The type of each tensor must match the model
+ * input element type and shape (except batch dimension). Total size of tensors must match the input size.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_tensors_by_const_port(ov_infer_request_t* infer_request, const ov_output_const_node_t* port, const ov_tensor_list_t* tensors);
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:123]: The same

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 13 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:108]: yes, you are right, such structure data can pass by value without any negative impact.

BTW, we re-define tensor list as below:
```
typedef struct {
    ov_tensor_t** tensors;
    size_t size;
} ov_tensor_list_t;
```
We can access each item by operator[] in C scope, each item itself is pointer.

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 15]
```diff
@@ -70,37 +71,149 @@ OPENVINO_C_API(ov_status_e)
 ov_infer_request_set_tensor(ov_infer_request_t* infer_request, const char* tensor_name, const ov_tensor_t* tensor);
 
 /**
- * @brief Sets an input tensor to infer on.
+ * @brief Sets an input/output tensor to infer on by the port of input/output tensor.
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:74]: Port it is not a part of tensor, it is a model part. I think this sentence can be changed to:
```suggestion
 * @brief Sets an input/output tensor to infer for the port.
```

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 16]
```diff
@@ -70,37 +71,149 @@ OPENVINO_C_API(ov_status_e)
...
  * @param tensor Reference to the tensor.
  * @return Status code of the operation: OK(0) for success.
  */
 OPENVINO_C_API(ov_status_e)
-ov_infer_request_set_input_tensor(ov_infer_request_t* infer_request, size_t idx, const ov_tensor_t* tensor);
+ov_infer_request_set_tensor_by_port(ov_infer_request_t* infer_request, const ov_output_node_t* port, const ov_tensor_t* tensor);
 
 /**
- * @brief Gets an input/output tensor to infer on.
+ * @brief Sets an input/output tensor to infer on by the const port of input/output tensor.
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:85]: The same comment that port it is not a part of tensor

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 13 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h]: As we aligned, C API will not support set tensor list for infer_request.

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 14 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h]: As we aligned, C API will not support set tensor list for infer_request.

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 17]
```diff
@@ -70,37 +71,149 @@ OPENVINO_C_API(ov_status_e)
...
+ * @param idx Index of the tensor to get. @p idx. If the tensor with the specified @p idx is not found, an exception
+ * @param tensor Reference to the tensor.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_get_output_tensor_by_index(const ov_infer_request_t* infer_request, const size_t idx, ov_tensor_t** tensor);
+
+/**
+ * @brief Gets an output tensor from the model with only one output tensor.
+ * @note f model has several outputs, an exception is thrown.
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:209]:
```suggestion
 * @note if model has several outputs, an exception is thrown.
```
I think we don't throw exceptions in C API

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 18]
```diff
@@ -70,37 +71,149 @@ OPENVINO_C_API(ov_status_e)
...
- * @param tensor_name  Name of the input or output tensor.
+ * @param tensor Reference to the tensor.
+ * @return Status code of the operation: OK(0) for success.
+ */
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_output_tensor(ov_infer_request_t* infer_request, const ov_tensor_t* tensor);
+
+/**
+ * @brief Gets an input/output tensor by the name of tensor.
+ * @note If model has several outputs, an exception is thrown.
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:142]: The same

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 19]
```diff
@@ -70,37 +71,149 @@ OPENVINO_C_API(ov_status_e)
...
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param tensor_name Name of the input or output tensor to get.
  * @param tensor Reference to the tensor.
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:146]: Pointer to tensor ptr?

**Review by ilyachur — COMMENTED** [review]
In general LGTM some comments for the documentation
@ilya-lavrenov Please also take a look

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 19 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:146]: ov_tensor_t is hidden structure, we should use ov_tensor_t * to access it, and this API will get a ov_tensor_t *, so we will set ov_tensor_t **.

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 17 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:209]: ok, will update.

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 18 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:142]: will update.

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 17 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:209]: yes, this exception will be caught by C API and will not expose to C user.

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 18 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h]: done

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 17 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h]: done

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 19 above)
Comment (ilyachur) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:154]: I mean that it is not a reference :)

File: src/bindings/c/include/openvino/c/ov_infer_request.h [hunk 20]
```diff
@@ -70,68 +71,207 @@ OPENVINO_C_API(ov_status_e)
...
+OPENVINO_C_API(ov_status_e)
+ov_infer_request_set_input_tensor(ov_infer_request_t* infer_request, const ov_tensor_t* tensor);
+
+/**
+ * @brief Set an output tensor to infer by the index of output tensor.
+ * @note Index of the output preserved accross ov_model_t, ov_compiled_model_t.
+ * @ingroup infer_request
+ * @param infer_request A pointer to the ov_infer_request_t.
+ * @param idx Index of the output tensor.
+ * @param tensor Reference to the tensor.
```
Comment (ilyachur) [inline src/bindings/c/include/openvino/c/ov_infer_request.h:130]: But I see reference in a lot of places. I think it can be fixed separately in next PR.

**Review by ilyachur — APPROVED** [review]

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 19 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:154]: yes, it is to return one

File: src/bindings/c/include/openvino/c/ov_infer_request.h (hunk 20 above)
Comment (riverlijunjie) [reply src/bindings/c/include/openvino/c/ov_infer_request.h:130]: There is some inappropriate description in documentation, it is input argument and pass by pointer.

**Pull request merged**